### PR TITLE
Streamlining HTML report code

### PR DESCRIPTION
**Pull request by standage**
The purpose of this branch is to clean up the code responsible for collating and rendering the HTML report for the end-to-end analysis pipeline.

----------

- [x] Changes are clearly described above
- [x] Any relevant issue threads are referenced in the description
- [x] Any new features are tested (see the [development manual](https://microhapulator.readthedocs.io/en/stable/devel.html) for details)
- [x] CLI documentation (see [docs/cli.md](docs/cli.md)) and Python API documentation (see [microhapulator/api.py](microhapulator/api.py)) are up-to-date and in sync
- [x] Substantial changes are documented in CHANGELOG.md (see https://keepachangelog.com/en/1.0.0/)

### CODE REVIEW

**Comment by standage** [comment]
I've drafted some classes for aggregating the data for the first QC section of the report in `qcstats.py` and `qcsummaries.py`. The following demonstrates how to access the variables needed to populate the report.

```python
>>> from microhapulator.qcsummary import PairedReadQCSummary, SingleEndReadQCSummary
>>> 
>>> qc = SingleEndReadQCSummary.collect(["SRM8398-1", "SRM8398-2", "SRM8398-3"], workdir="scratch/WD_clean2/")
>>> for sample, stats in qc.items():
...   print(sample, stats.total_reads, stats.filtered_ambig, stats.filtered_length, stats.retention, sep="\t")
... 
SRM8398-1       211,819 106,476 (50.3%) 6,092 (2.9%)    99,251 (46.9%)
SRM8398-2       328,618 180,940 (55.1%) 7,471 (2.3%)    140,207 (42.7%)
SRM8398-3       245,967 107,726 (43.8%) 8,225 (3.3%)    130,016 (52.9%)
>>> 
>>> 
>>> qc = PairedReadQCSummary.collect(["SRM8398-1", "SRM8398-2", "SRM8398-3"], workdir="scratch/WD_nimagen_testC")
>>> for sample, stats in qc.items():
...     print(sample, stats.ambig.total_reads, stats.ambig.excluded_r1, stats.ambig.excluded_r2, stats.ambig.excluded_both, stats.ambig.excluded, stats.ambig.retained, stats.ambig.retention_rate, sep="\t")
... 
SRM8398-1       134,086 827     477     42,698  44,002  90084   67.2%
SRM8398-2       208,872 1,152   1,767   77,869  80,788  128084  61.3%
SRM8398-3       167,704 883     462     45,617  46,962  120742  72.0%
>>> 
>>> for sample, stats in qc.items():
...   print(sample, stats.merge.total_reads, stats.merge.merged_reads, stats.merge.merge_rate, sep="\t")
... 
SRM8398-1       90,084  89,099  98.9%
SRM8398-2       128,084 126,633 98.9%
SRM8398-3       120,742 119,772 99.2%
>>> 
>>> for sample, stats in qc.items():
...   print(sample, stats.length.total_reads, stats.length.excluded, stats.length.kept, stats.length.retention_rate, sep="\t")
... 
SRM8398-1       89,099  2,452   86,647  97.2%
SRM8398-2       126,633 3,298   123,335 97.4%
SRM8398-3       119,772 3,546   116,226 97.0%
>>>
```

File: microhapulator/qcsummary.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+# -------------------------------------------------------------------------------------------------
+
+from .qcstats import FilterStats, PairedAmbiguityFilterStats, ReadMergingStats
+from dataclasses import dataclass
+
+
+@dataclass
+class SingleEndReadQCSummary:
+    ambig: FilterStats
+    length: FilterStats
```
Comment (standage) [inline microhapulator/qcsummary.py:20]: The table of QC stats for single-end data can be managed by a couple of simple `FilterStats` data structures, one for ambiguity filtering and another for length filtering. A few methods are defined on the class to make sure the right denominator is used for calculating the percentage of the total reads being filtered at each step.

File: microhapulator/qcsummary.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        for sample in sorted(samples):
+            sampleqc[sample] = SingleEndReadQCSummary.from_workdir(workdir, sample)
+        return sampleqc
+
+
+@dataclass
+class PairedReadQCSummary:
+    ambig: PairedAmbiguityFilterStats
+    merge: ReadMergingStats
+    length: FilterStats
```
Comment (standage) [inline microhapulator/qcsummary.py:67]: For paired-end data, the simple `FilterStats` data structure works for length filtering post-merge, but the ambiguity filtering and read merge stats needed their own dedicated data structures. This class is essentially a container for the three data structures, along with convenience methods for loading from the working directory.

File: microhapulator/qcstats.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+from dataclasses import dataclass
+import pandas as pd
+import re
+
+
+@dataclass
+class FilterStats:
+    filtered: int
+    retained: int
```
Comment (standage) [inline microhapulator/qcstats.py:21]: Simple class for reads kept vs discarded.

File: microhapulator/qcstats.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        return stats
+
+
+@dataclass
+class PairedAmbiguityFilterStats:
+    total: int
+    failed_r1: int
+    failed_r2: int
+    failed_both: int
+    retained: int
```
Comment (standage) [inline microhapulator/qcstats.py:59]: Ambiguity filtering stats, keeping track of which fragment(s) are responsible for the pair being discarded.

File: microhapulator/qcstats.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        return stats
+
+
+@dataclass
+class ReadMergingStats:
+    total: int
+    combined: int
+    innies: int
+    outies: int
+    uncombined: int
```
Comment (standage) [inline microhapulator/qcstats.py:112]: Merging statistics.

File: CHANGELOG.md [hunk 6]
```diff
@@ -17,6 +17,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
     - Replaced read QC donut plots with a stacked bar chart (#168).
     - Replaced FastQC report links with MultiQC link (#169).
     - Revised report text, added figure captions and table titles (#172).
+    - Streamlined the Python code responsible for generating the report (#175).
```
Comment (standage) [inline CHANGELOG.md:20]: If we end up splitting these updates across multiple PRs, as I'd suggest, then we can just add all of the relevant PRs to this change log entry.

File: microhapulator/data/template.html [hunk 7]
```diff
@@ -187,16 +187,16 @@ <h2>Read QA/QC</h2>
...
-                    <td class="alnrt">{{"{:,}".format(ambig_reads_df.iloc[0]["PairsRemoved"])}}</td>
-                    <td class="alnrt">{{"{:,}".format(ambig_reads_df.iloc[0]["PairsKept"])}}</td>
-                    <td class="alnrt">{{"{:.2f}".format(ambig_reads_df.iloc[0]["PairsKept"] / ambig_reads_df.iloc[0]["TotalPairs"] * 100)}}%</td>
+                    <td class="alnrt">{{stats.ambig.total_reads}}</td>
+                    <td class="alnrt">{{stats.ambig.excluded_r1}}</td>
+                    <td class="alnrt">{{stats.ambig.excluded_r2}}</td>
+                    <td class="alnrt">{{stats.ambig.excluded_both}}</td>
+                    <td class="alnrt">{{stats.ambig.excluded}}</td>
+                    <td class="alnrt">{{stats.ambig.retained}}</td>
+                    <td class="alnrt">{{stats.ambig.retention_rate}}</td>
```
Comment (standage) [inline microhapulator/data/template.html:199]: Modest improvement in clarity here.

File: microhapulator/data/template.html [hunk 8]
```diff
@@ -216,13 +216,13 @@ <h2>Read QA/QC</h2>
...
-                    <td class="alnrt">{{"{:,}".format(ambig_reads_df.iloc[0]["ReadsRemoved"] + ambig_reads_df.iloc[0]["ReadsKept"])}}</td>
-                    <td class="alnrt">{{"{:,} ({:.1f}%)".format(ambig_reads_df.iloc[0]["ReadsRemoved"], ambig_reads_df.iloc[0]["ReadsRemoved"] / (ambig_reads_df.iloc[0]["ReadsRemoved"] + ambig_reads_df.iloc[0]["ReadsKept"]) * 100)}}</td>
-                    <td class="alnrt">{{"{:,} ({:.1f}%)".format(summary[summary.Sample == sample].iloc[0].LengthFailed, summary[summary.Sample == sample].iloc[0].LengthFailed / (ambig_reads_df.iloc[0]["ReadsRemoved"] + ambig_reads_df.iloc[0]["ReadsKept"]) * 100)}}</td>
-                    <td class="alnrt">{{"{:,} ({:.1f}%)".format(summary[summary.Sample == sample].iloc[0].LengthPassed, summary[summary.Sample == sample].iloc[0].LengthPassed / (ambig_reads_df.iloc[0]["ReadsRemoved"] + ambig_reads_df.iloc[0]["ReadsKept"]) * 100)}}</td>
+                    <td class="alnrt">{{stats.total_reads}}</td>
+                    <td class="alnrt">{{stats.filtered_ambig}}</td>
+                    <td class="alnrt">{{stats.filtered_length}}</td>
+                    <td class="alnrt">{{stats.retention}}</td>
```
Comment (standage) [inline microhapulator/data/template.html:225]: Considerable improvement in clarity here.

File: microhapulator/pipeaux.py [hunk 9]
```diff
@@ -200,16 +200,6 @@ def read_length_table_single_end(samples):
...
-def parse_ambig_reads(samples):
-    ambig_read_counts_all = dict()
-    for sample in samples:
-        ambig_reads_df = pd.read_csv(
-            f"analysis/{sample}/{sample}-ambig-read-counts.txt", sep="\t", index_col=None
-        )
-        ambig_read_counts_all[sample] = ambig_reads_df
-    return ambig_read_counts_all
-
-
```
Comment (standage) [inline microhapulator/pipeaux.py:212]: Obsolete function.

File: microhapulator/pipeaux.py [hunk 10]
```diff
@@ -34,6 +34,7 @@ def full_reference_index_files(fasta):
 
 
 def parse_flash_summary(logfile):
+    # FIXME delete once this refactor is through
     with open(logfile, "r") as fh:
         data = fh.read()
         tp = re.search(r"Total pairs:\s+(\d+)", data).group(1)
```
Comment (standage) [inline microhapulator/pipeaux.py:40]: This is also an obsolete function, but it's being used by the `aggregate_summary` function. So this comment is a reminder to take care of it.

**Comment by agshumate** [comment]
The code definitely looks a lot cleaner and functional review looks good too. One note is that in the future, if we implement other filters for paired reads, we should generalized  the ` PairedAmbiguityFilterStats` class to just ` PairedFilterStats` as we will want to track the same stats for any filtering that we do. But i don't think we need to worry about that yet since currently the only filter was have for paired-end data is ambiguity filtering.

**Pull request merged**